File: Tutorials/WPF/Classic/CS/DataAccess/Order.cs
using DevExpress.Xpo;
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace XpoTutorial {

    public class Order : BaseObject, IEditableObject {
        public Order(Session session) : base(session) { }
        string productName;
        public string ProductName {
            get { return productName; }
            set { SetPropertyValue(nameof(ProductName), ref productName, value); }
        }
        DateTime orderDate;
        public DateTime OrderDate {
            get { return orderDate; }
            set { SetPropertyValue(nameof(OrderDate), ref orderDate, value); }
        }
        decimal? freight;
        public decimal? Freight {
            get { return freight; }
            set { SetPropertyValue(nameof(Freight), ref freight, value); }
        }
        Customer customer;
        [Association("CustomerOrders")]
        public Customer Customer {
            get { return customer; }
            set { SetPropertyValue(nameof(Customer), ref customer, value); }
        }

        #region IEditableObject implementation
        
        bool isEditing;
        string oldProductName;
        DateTime oldOrderDate;
        decimal? oldFreight;
        Customer oldCustomer;

        public void BeginEdit() {
            if(isEditing) {
                throw new InvalidOperationException();
            }
            oldProductName = ProductName;
            oldOrderDate = OrderDate;
            oldFreight = Freight;
            oldCustomer = Customer;
            isEditing = true;
        }

        public void CancelEdit() {
            if(!isEditing) {
                throw new InvalidOperationException();
            }
            ProductName = oldProductName;
            OrderDate = oldOrderDate;
            Freight = oldFreight;
            Customer = oldCustomer;
            isEditing = false;
        }

        public void EndEdit() {
            if(!isEditing) {
                throw new InvalidOperationException();
            }
            isEditing = false;
        }

        #endregion
    }
}

File: Tutorials/WPF/Classic/VB/DataAccess/Order.vb
Imports DevExpress.Xpo
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel

Namespace XpoTutorial

	Public Class Order
		Inherits BaseObject
		Implements IEditableObject

		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Private fProductName As String
		Public Property ProductName() As String
			Get
				Return fProductName
			End Get
			Set(ByVal value As String)
				SetPropertyValue(NameOf(ProductName), fProductName, value)
			End Set
		End Property
		Private fOrderDate As Date
		Public Property OrderDate() As Date
			Get
				Return fOrderDate
			End Get
			Set(ByVal value As Date)
				SetPropertyValue(NameOf(OrderDate), fOrderDate, value)
			End Set
		End Property
        Private fFreight As Decimal?
        Public Property Freight() As Decimal?
			Get
				Return fFreight
			End Get
			Set(ByVal value? As Decimal)
				SetPropertyValue(NameOf(Freight), fFreight, value)
			End Set
		End Property
		Private fCustomer As Customer
		<Association("CustomerOrders")>
		Public Property Customer() As Customer
			Get
				Return fCustomer
			End Get
			Set(ByVal value As Customer)
				SetPropertyValue(NameOf(Customer), fCustomer, value)
			End Set
		End Property

		#Region "IEditableObject implementation"

		Private isEditing As Boolean
		Private oldProductName As String
		Private oldOrderDate As Date
		Private oldFreight? As Decimal
		Private oldCustomer As Customer

		Public Sub BeginEdit() Implements IEditableObject.BeginEdit
			If isEditing Then
				Throw New InvalidOperationException()
			End If
			oldProductName = ProductName
			oldOrderDate = OrderDate
			oldFreight = Freight
			oldCustomer = Customer
			isEditing = True
		End Sub

		Public Sub CancelEdit() Implements IEditableObject.CancelEdit
			If Not isEditing Then
				Throw New InvalidOperationException()
			End If
			ProductName = oldProductName
			OrderDate = oldOrderDate
			Freight = oldFreight
			Customer = oldCustomer
			isEditing = False
		End Sub

		Public Sub EndEdit() Implements IEditableObject.EndEdit
			If Not isEditing Then
				Throw New InvalidOperationException()
			End If
			isEditing = False
		End Sub

		#End Region
	End Class
End Namespace
